### PR TITLE
dnsproxy: update dnsproxy benchmark memory calculation

### PR DESCRIPTION
**Pull request by odinuge**
**TL;DR:** After a lot of testing, and some very strange results, I realized that the memory printouts were a bit less deterministic than expected. This should clean that up, and make is easier to actually compare the heap use of different approaches.

Previously, the printed memory values depended greatly on both the size
of the loaded CNPs in memory, as well as when the gc ran for the last
time. That favored solutions that allocated a lot of memory in the start
of the test (but still inside the actual testing section), while
solution allocating a lot of memory in the end, even if they removed all
the referneces to it, were punished due the fact that we printed the
memory stats before forcing a gc.

For heap, the diff between the "Before Setup" and "After Test" should be
inspected, since "Before Test" has all the CNPs in memory. When we hit
"After test", those CNPs have been garbage collected.

New output:
```
root@10fc3f693c5d:/code# go test -tags privileged_tests -v -run '^$' -bench Benchmark_perEPAllow_setPortRulesForID_large -benchmem -benchtime 1x ./pkg/fqdn/dnsproxy
goos: linux
goarch: amd64
pkg: github.com/cilium/cilium/pkg/fqdn/dnsproxy
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_perEPAllow_setPortRulesForID_large
Before Setup (N=1,EPs=20,cache=128)
Alloc = 3 MiB   HeapInuse = 5 MiB       Sys = 20 MiB    NumGC = 6
Before Test (N=1,EPs=20,cache=128)
Alloc = 32 MiB  HeapInuse = 50 MiB      Sys = 987 MiB   NumGC = 16
After Test (N=1,EPs=20,cache=128)
Alloc = 341 MiB HeapInuse = 374 MiB     Sys = 987 MiB   NumGC = 38
Before Setup (N=1,EPs=20,cache=128)
Alloc = 3 MiB   HeapInuse = 6 MiB       Sys = 987 MiB   NumGC = 40
Before Test (N=1,EPs=20,cache=128)
Alloc = 32 MiB  HeapInuse = 56 MiB      Sys = 1020 MiB  NumGC = 49
After Test (N=1,EPs=20,cache=128)
Alloc = 344 MiB HeapInuse = 378 MiB     Sys = 1020 MiB  NumGC = 71
Benchmark_perEPAllow_setPortRulesForID_large-7                 1        4301736800 ns/op        3047525016 B/op 26526920 allocs/op
PASS
ok      github.com/cilium/cilium/pkg/fqdn/dnsproxy      20.796s
```

Fixes: https://github.com/cilium/cilium/pull/19383
Fixes: https://github.com/cilium/cilium/commit/38c00367c7e0cefa1c9bbd51ad4414a839015bff ("dnsproxy: Add benchmark for large FQDN-based CNPs")
Signed-off-by: Odin Ugedal <ougedal@palantir.com>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Thanks for contributing!

Fixes: #issue-number

### CODE REVIEW

File: pkg/fqdn/dnsproxy/proxy_test.go [hunk 1]
```diff
@@ -1126,30 +1141,36 @@ func Benchmark_perEPAllow_setPortRulesForID_large(b *testing.B) {
 		}
 	}
 
-	fmt.Printf("Before (N=%v)\n", b.N)
-
 	runtime.GC()
+	m = getMemStats()
+	fmt.Printf("Before Test (N=%v,EPs=%d,cache=%d)\n", b.N, numEPs, cacheSize)
```
Comment (odinuge) [inline pkg/fqdn/dnsproxy/proxy_test.go:1146]: We can skip this one to avoid confusion, but I thought the GC count could be interesting, since it will be higher here that in the very beginning.

File: pkg/fqdn/dnsproxy/proxy_test.go [hunk 2]
```diff
@@ -1068,8 +1068,23 @@ func Benchmark_perEPAllow_setPortRulesForID(b *testing.B) {
 
 func Benchmark_perEPAllow_setPortRulesForID_large(b *testing.B) {
 	b.Skip()
+	cacheSize := 1024
+	numEPs := uint64(40)
```
Comment (odinuge) [inline pkg/fqdn/dnsproxy/proxy_test.go]: OOo, I now see that I doubled it to `40` here. Happy to dial it down again if you think 20 is more suitable.

Comment (christarazi) [reply pkg/fqdn/dnsproxy/proxy_test.go]: Let's keep 20 so we have a consistent baseline from which we were testing.

**Review by christarazi — APPROVED** [review]

**Comment by christarazi** [comment]
Could you also add the PR and commit that this PR is fixing up to this PR's description? This is useful to know how far back we need to backport these improvements to.

File: pkg/fqdn/dnsproxy/proxy_test.go (hunk 2 above)
Comment (odinuge) [reply pkg/fqdn/dnsproxy/proxy_test.go]: Agree! Keeping the cache size and the numEPs the same as before!

**Comment by odinuge** [comment]
> Could you also add the PR and commit that this PR is fixing up to this PR's description? This is useful to know how far back we need to backport these improvements to.

Yeah, ofc.! 😄 Pushed a new version including that now!

File: pkg/fqdn/dnsproxy/proxy_test.go (hunk 1 above)
Comment (christarazi) [reply pkg/fqdn/dnsproxy/proxy_test.go:1146]: Yeah let's skip it.

**Comment by christarazi** [comment]
Aside from addressing https://github.com/cilium/cilium/pull/20305#discussion_r906186477, this PR should be good to merge considering that we don't need to run CI against it since it's just updating the benchmark testing code.

**Comment by gandro** [comment]
This has been marked for backport to v1.10, but seems to depend on https://github.com/cilium/cilium/pull/19632 - which has not been backported to v1.10. What's the preferred way forward? @christarazi

**Comment by christarazi** [comment]
I've removed it from v1.10.
